constructors.go

```diff
@@ -0,0 +1,167 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+)
+
+type scopedProvider struct {
+	Provider
+
+	prefix string
+}
+
+var _ Provider = (*scopedProvider)(nil)
+
+func (s *scopedProvider) Get(key string) Value {
+	return s.Provider.Get(fmt.Sprintf("%s.%s", s.prefix, key))
+}
+
+// NewScopedProvider wraps a provider and adds a prefix to all Get calls.
+func NewScopedProvider(prefix string, provider Provider) Provider {
+	if prefix == "" {
+		return provider
+	}
+	return &scopedProvider{provider, prefix}
+}
+
+// NewProviderGroup composes multiple providers, with later providers
+// overriding earlier ones. The merge logic is described in the package-level
+// documentation. To preserve backward compatibility, the resulting provider
+// disables strict unmarshalling.
+//
+// Deprecated: compose multiple sources of configuration using NewYAML and the
+// Static, File, and/or Source options directly. This enables strict
+// unmarshalling by default and allows use of other options at the same time.
+func NewProviderGroup(name string, providers ...Provider) (Provider, error) {
+	opts := make([]YAMLOption, 0, len(providers)+2)
+	opts = append(opts, Name(name), Permissive())
+	for _, p := range providers {
+		opts = append(opts, Static(p.Get(Root).Value()))
+	}
+	return NewYAML(opts...)
+}
+
+// NewStaticProvider serializes a Go data structure to YAML, then loads it
+// into a provider. To preserve backward compatibility, the resulting provider
+// disables strict unmarshalling.
+//
+// Deprecated: use NewYAML and the Static option directly. This enables strict
+// unmarshalling by default and allows use of other options at the same time.
+func NewStaticProvider(data interface{}) (Provider, error) {
+	return NewStaticProviderWithExpand(data, nil)
+}
+
+// NewStaticProviderWithExpand serializes a Go data structure to YAML, expands
+// any environment variable references using the supplied lookup function,
+// then loads the result into a provider. See the Expand option for a
+// description of the environment variable replacement syntax. To preserve
+// backward compatibility, the resulting provider disables strict
+// unmarshalling.
+//
+// Deprecated: use NewYAML and the Static and Expand options directly. This
+// enables strict unmarshalling by default and allows use of other options at
+// the same time.
+func NewStaticProviderWithExpand(data interface{}, lookup LookupFunc) (Provider, error) {
+	return NewYAML(Static(data), Expand(lookup), Permissive())
+}
+
+// NewYAMLProviderFromBytes merges multiple YAML-formatted byte slices into a
+// single provider. As always, later configuration blobs override earlier ones
```

Comment:
nit: Consider dropping `As always` since the other functions don't include it and this might be the first function the reader is viewing.

Reply:
Good point, will do.